pkg/tnf/operator/starter.go

```diff
@@ -138,19 +128,84 @@ func HandleDualReplicaClusters(ctx context.Context,
+		if !ok {
+			klog.Warningf("failed to convert added object to Node %+v", obj)
+			return
+		}
+		klog.Infof("node added: %s", node.GetName())
+
+		// ensure we have both control plane nodes before creating jobs
+		nodeList, err := controlPlaneNodeLister.List(labels.Everything())
+		if err != nil {
+			klog.Errorf("failed to list control plane nodes while waiting to create TNF jobs: %v", err)
+			return
+		}
+		if len(nodeList) != 2 {
+			klog.Info("not starting TNF jobs yet, waiting for 2 control plane nodes to exist")
+			return
+		}
+		klog.Infof("found 2 control plane nodes (%q, %q)", nodeList[0].GetName(), nodeList[1].GetName())
+
+		// we can have 2 nodes on the first call of AddFunc already, ensure we create job controllers once only
+		once.Do(func() {
+			// this potentially needs some time when we wait for etcd bootstrap to complete, so run it in a goroutine,
+			// to not block the event handler
+			go h.handleNodes(nodeList, ctx, controllerContext, operatorClient, kubeClient, kubeInformersForNamespaces)
+		})
+	}
+}
+
+func (h *DualReplicaClusterHandler) handleNodes(
+	nodeList []*corev1.Node,
+	ctx context.Context,
+	controllerContext *controllercmd.ControllerContext,
+	operatorClient v1helpers.StaticPodOperatorClient,
+	kubeClient kubernetes.Interface,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces) {
+
+	if err := h.waitForEtcdBootstrapCompleted(ctx, operatorClient); err != nil {
+		klog.Errorf("failed to wait for etcd bootstrap: %v", err)
+		return
 	}
-	// dual replica currently topology has to be enabled by a feature gate
-	if enabledDualReplicaFeature, err := dualreplicahelpers.DualReplicaFeatureGateEnabled(featureGateAccessor); err != nil {
-		return false, fmt.Errorf("could not determine DualReplicaFeatureGateEnabled, aborting controller start: %w", err)
-	} else if !enabledDualReplicaFeature {
-		return false, fmt.Errorf("detected dual replica topology, but dual replica feature gate is disabled, aborting controller start")
+	klog.Infof("bootstrap completed, creating TNF job controllers")
+
+	// the order of job creation does not matter, the jobs wait on each other as needed
+	for _, node := range nodeList {
+		runJobController(ctx, tools.JobTypeAuth, &node.Name, controllerContext, operatorClient, kubeClient, kubeInformersForNamespaces)
+		runJobController(ctx, tools.JobTypeAfterSetup, &node.Name, controllerContext, operatorClient, kubeClient, kubeInformersForNamespaces)
 	}
-	return true, nil
+	runJobController(ctx, tools.JobTypeSetup, nil, controllerContext, operatorClient, kubeClient, kubeInformersForNamespaces)
+	runJobController(ctx, tools.JobTypeFencing, nil, controllerContext, operatorClient, kubeClient, kubeInformersForNamespaces)
+}
+
+func (h *DualReplicaClusterHandler) waitForEtcdBootstrapCompleted(ctx context.Context, operatorClient v1helpers.StaticPodOperatorClient) error {
+	isEtcdRunningInCluster, err := ceohelpers.IsEtcdRunningInCluster(ctx, operatorClient)
```

Comment:
Same question about whether we need to sync before this call

Reply:
yeah, so you need to use the informer and lister facility if you want cache coherence. 

If you're using a client, then it's basically a synchronous API call every time, so there is no caching involved.